### ISSUE TITLE
Enabling persistent user instance systemd

### ISSUE THREAD

**Issue by CMCDragonkai** [NixOS/nixpkgs#3702]
Apparently user specific systemd instances only exist while there is a session for a specific user. If the last session exits, then the user specific systemd instance exits as well.

There's a way to make a user specific systemd instance persistent, so that users can run unattended long running services: https://wiki.archlinux.org/index.php/Systemd/User#Automatic_start-up_of_systemd_user_instances

Basically you need to activate this option: http://www.freedesktop.org/software/systemd/man/loginctl.html#enable-linger%20USER...

```
loginctl enable-linger username
```

However this option is not available in NixOS or NixPkgs, I searched the repo. Also the logind extraConfig doesn't cover it.

I think there might need to be a new option as part of the user management that makes the user instance systemd persistent, and run the above command on bootup and the opposite if the setting changes?

It should of course only run once if it has already ran, even though it is idempotent. I wonder how you can do this?

Right now I'm simply running it during when I pack the image.

**Comment by layus** [comment]
There seems to be no option in logind.conf to enable this.
Basically, lingering is enabled in systemd at boot if it finds a file /var/lib/systemd/linger/<username>.

We could use a `system.activationScript.loginctl-linger` script to call `loginctl enable-linger <username>` for each user in the list for which `logind show-user <username>` does not return "Linger=yes".

This would work just like user management in https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/config/users-groups.nix#L504-510, except that that script handles much more cases, and optionally handles the deletion of unspecified users.

To go further, you could too remove lingering for users not in the list, but this requires more work and should depend on some option equivalent to `users.mutableUsers` or you may surprise some users.

**Comment by michalrus** [comment]
This bit me today, as I *forgot* to **MANUALLY** run `sudo loginctl enable-linger m` on a new machine running *declarative* NixOS… :confused: And my multiplexer was killed.

@layus, nb., one should not link to `master`, as `master` is mutable.

**Comment by michalrus** [comment]
If anyone’s interested, right now, you can use something along the lines of https://github.com/michalrus/dotfiles/commit/455c18a81c1fcaaccb99625c2f88402ebd08ac23.

**Comment by Shados** [comment]
Actually, you can effectively 'enable lingering globally' with `KillUserProcesses=no` in `logind.conf`.

That said, it would probably be better to just compile systemd `--without-kill-user-processes` to disable this behaviour by default, as this is already being done in e.g. Arch and Debian (I think?). Killing daemonized user processes on logout breaks user expectations pretty hugely, and is really only beneficial on deliberately very locked down systems (e.g. lan cafe PCs, kiosks), where it can be manually enabled if required. On normal systems, if there are non-daemon user programs that aren't terminating on receiving `SIGHUP`, that's a bug in said programs.

On a more general note, I'm not sure why systemd needs a new way for a program to express the desire to run beyond the lifespan of the current session, in the form of `systemd-run`. We already have `daemon()`.

**Comment by bjornfor** [comment]
FYI, NixOS has ```KillUserProcesses=no``` in ```logind.conf``` since 93c160fabc0e3b5fb3d997c840ef86a68f4bce00.

**Comment by layus** [comment]
@CMCDragonkai As bjornfor stated, this issue is now solved in master. Would you mind closing this issue ?

**Comment by CMCDragonkai** [comment]
I see the solution provided, but that now always forces lingering. The whole point of this is so that its optional and can be specific to the user session instead of a global setting. Anyway if that global setting, how would people turn off it now?

**Comment by layus** [comment]
As I understand it, systemd requires services tout declare somehow their intent to linger after user sessions. The current fix is not a solution. It is required to give some time to upstream packages (like gnu screen where the discussion is tough) to adapt to systemd. Afterwards, we will probably return to systemd default.

For override, see comment in https://github.com/NixOS/nixpkgs/commit/93c160fabc0e3b5fb3d997c840ef86a68f4bce00#commitcomment-18503557

That being said, my personal opinion is that user processes should terminate nicely when the user logs out. It feels like systemd trying to fix the task of the login manager here.

**Comment by CMCDragonkai** [comment]
@globin Can you reopen?

**Comment by michalrus** [comment]
> If anyone’s interested, right now, you can use something along the lines of https://github.com/michalrus/dotfiles/commit/455c18a81c1fcaaccb99625c2f88402ebd08ac23.

A bit more polished version (now you can do `users.users.somebody.linger = true` and disabling/enabling will work): https://github.com/michalrus/dotfiles/commit/ebd5fa9583f82589f23531647aa677feb3f8d344#diff-4d353005ef5b3e37f33c07332b8523edR1

**Comment by CMCDragonkai** [comment]
Is it possible to make the setting immutable after it is set using your method @michalrus

**Comment by michalrus** [comment]
@CMCDragonkai, what do you mean by immutable? In your `configuration.nix`, after adding my `loginctl-linger.nix` to `imports = []`, you can do:

```nix
{
  imports = [
    (pkgs.fetchurl { src = "https://raw.githubusercontent.com/michalrus/dotfiles/ebd5fa9583f82589f23531647aa677feb3f8d344/.nixos-config.symlink/modules/loginctl-linger.nix"; sha256 = "…"; })
  ];
  users.extraUsers.cmcdragonkai = {
    description = "Roger Qiu";
    isNormalUser = true;
    linger = true;
  };
}
```

And this is quite immutable. :smiley: 

Or do you mean immutable wrt. manually running `sudo loginctl enable-linger cmcdragonkai`? =P That would require patching `loginctl`. **BUT**. If you mess up manually, and rerun `sudo nixos-rebuild switch`, all will be fixed.

N.b., I’ve opened a feature quest in systemd for the functionality of listing users with lingering enabled → https://github.com/systemd/systemd/issues/3907.

**Comment by CMCDragonkai** [comment]
Yep, I was interested in whether the setting can be made immutable in the live environment thus preventing configuration drift. (I usually make /etc readonly). But this is a better solution than the previous one. Is it possible to simplify to just an activation script? A PR to nixpkgs proper would be great.

**Comment by michalrus** [comment]
@CMCDragonkai, sure, I was thinking where that code should go… and dunno.

As for making lingering immutable, I don’t know. The simple and good way would be to patch systemd to check whether the user is allowed to linger **not** in `/var/lib/systemd/linger/`, but somewhere in the `/nix/store/`. Then `sudo loginctl enable-linger someone` would fail on the syscall level, as the `/nix/store/` location would be immutable.

@joachifm, @globin, do you think my current solution is enough for nixpkgs? It’s in https://github.com/michalrus/dotfiles/commit/ebd5fa9583f82589f23531647aa677feb3f8d344#diff-4d353005ef5b3e37f33c07332b8523edR1 I’d rather have patched systemd, but all mentions of `/var/lib/systemd/linger` should be replaced in systemd’s runtime with some configurable directory in `/nix/store`. Makes it more complicated…

WDYT?

**Comment by CMCDragonkai** [comment]
If we do your loginctl PR, then I think this commit should be reversed in the same PR, as otherwise users are lingering by default anyway.

**Comment by CMCDragonkai** [comment]
https://github.com/NixOS/nixpkgs/commit/93c160fabc0e3b5fb3d997c840ef86a68f4bce00#commitcomment-18503557

Forgot to add the commit.

**Comment by CMCDragonkai** [comment]
Ping @globin this issue is not solved. The best solution that is available at the moment is @michalrus script: https://github.com/michalrus/dotfiles/commit/ebd5fa9583f82589f23531647aa677feb3f8d344#diff-4d353005ef5b3e37f33c07332b8523edR1 

The better future solution is to for systemd to make this declaratively configurable.

**Comment by Profpatsch** [comment]
(triage) what’s the status?

**Comment by Baughn** [comment]
This is still a problem.

We could fix it in a variety of ways, for example by creating files in /var/lib/systemd/linger based on a user setting, but that would have to be a tristate; by default users can create/remove those using `loginctl (enable|disable)-linger`.

**Comment by Profpatsch** [comment]
It shouldn’t be up to the user whether to enable lingering. We can remove a lot of complexity if we only allow this from declarative user configs (aka just generate the linger directory and symlink it based on the config).

**Comment by Profpatsch** [comment]
If you only allow declarative, a simple implementation like I used in our config should be fine:

```
    # This enables “lingering” for the CI user.
    # Inspired by the discussion (and linked code)
    # in https://github.com/NixOS/nixpkgs/issues/3702
    # This should just be a NixOS option really.
    system.activationScripts = {
      enableLingering = ''
        # remove all existing lingering users
        rm -r /var/lib/systemd/linger
        mkdir /var/lib/systemd/linger
        # enable for the subset of declared users
        touch /var/lib/systemd/linger/${constants.ci-user-name}
      '';
    };
```

**Comment by CMCDragonkai** [comment]
It could be a global policy, but then specific to each to each user?

**Comment by jonas-schievink** [comment]
Still an issue

**Comment by jonas-schievink** [comment]
Still an issue

**Comment by pshirshov** [comment]
@flokli @Kloenk @andir @edolstra : sorry about pinging you, but would it be possible to add a setting for this? Much appreciated!

**Comment by NorfairKing** [comment]
Ping because it's still relevant.

**Comment by michalrus** [comment]
Maybe something similar to ↓ would feel cleaner (no lingering processes, everything still under systemd’s control) / acceptable to some:

```nix
{
  systemd.services.my-gnu-screen = let
    me = "NorfairKing"; 
    constantSessionName = "7260c3b2-2e3d-4b22-8f8b-ab87de790446";
  in {
    serviceConfig = {
      User = me;
      Type = "forking";
      Restart = "always";
    };
    wantedBy = [ "multi-user.target" ];
    script = ''
      ${lib.optionalString (me == "root") "export SHELL=${config.users.defaultUserShell}"}
      source ${config.system.build.setEnvironment}
      cd "$HOME"
      exec "$SHELL" --login -c "exec ${pkgs.screen}/bin/screen -S ${constantSessionName} -d -m"
    '';    
  };
}
```

**Comment by claes** [comment]
> It shouldn’t be up to the user whether to enable lingering, it’s system/root policy. We can remove a lot of complexity if we only allow this from declarative user configs (aka just generate the linger directory and symlink it based on the config).

I think in logind's design it is the options KillUserProcesses, KillOnlyUsers and KillExcludeUsers which are system / root policy. A regular user is able to enable lingering for him/herself using loginctl enable-linger

**Comment by busti** [comment]
Is there anything major that actually speaks against adding this option to `users.users.<name>`?

**Comment by colemickens** [comment]
@graham33 I am trying to clean up my nixpkgs and this is one of the commits I've been carrying for some time. I think we should get something in and iterate if need be. I suspect a number of people are carrying some variant of this anyway.

If you send a PR, I'll review it!

**Comment by aanderse** [comment]
That would be great. ❤️

**Comment by thefossguy** [comment]
I'm waiting for an "official solution" (something like `users.users.<name>.linger = <bool>;`). But in the meantime, I am using the following config:

```diff
+  # temporary hack until official lingering support is added to `users.users`
+  systemd.tmpfiles.rules = [
+    "f /var/lib/systemd/linger/pratham"
+  ];
+
```

Keeping it here for future NixOS newbies like me, but hopefully this has been resolved by then. Do check the `users.users` module on the stable NixOS before you use this!

**Comment by ToxicFrog** [comment]
> It looks like @michalrus subsequently stopped using his module and deleted it. I updated it slightly to work with recent NixOS: https://gist.github.com/graham33/fdbdcc18317a621d9dd54beb36be6683
> 
> I agree with @busti that this seems like it's good enough to just add as a regular user option. Happy to make a PR if that helps?

I tried tidying this up and making it PR-able, but when testing it locally, the `loginctl enable-linger` commands fail with `read-only file system`. Looking at `strace` on logind, it's getting EROFS when it tries to modify the contents of `/var/lib/systemd/linger`.

Looking at the unit file for `systemd-logind.service`, it doesn't have that directory listed under `ReadWritePaths`, but it does have it listed as the `StateDirectory`, so I would expect it to be mounted writeable regardless.

I can rework the PR to instead just `touch` or `unlink` the linger files, but (a) that's less clean and (b) I'm not sure if `logind` does more than just manipulating those files (e.g. maybe it pokes `systemd` to tell it to start/stop the user slice as well or someting). 

How did you work around this?

**Comment by layus** [comment]
I cannot believe this is now fixed :party: !
Thanks to everybody involved over the years, this is amazing !

**Comment by CMCDragonkai** [comment]
Hooray.

**Comment by dschrempf** [comment]
It is amazing that this option exists, thank you!

I noticed, however, that it is not persistent when using a temporal root file system. Other options related to users are persistent. Is this behavior desired? Can it even be changed?

I can restore `/var/lib/systemd/linger` on boot, and all is fine, but just wanted to through in this thought. Cheers!

**Comment by colemickens** [comment]
Given that this was achieved via an activation script, I don't understand why this would matter?

**Comment by ToxicFrog** [comment]
Yeah, it shouldn't matter if `/var/lib/systemd/linger` persists -- it'll get recreated on activate, and `systemctl enable-linger` should also make sure the user's slice starts. So maybe it doesn't start _exactly_ on boot but it should start on activate at the latest.

Something that I think may be worth addressing, raised by SuperSandro2000 [here](https://github.com/NixOS/nixpkgs/pull/260248/files/e648d4646589c380ff3a9fb57824eacef1cee412#r1359637530), is that this change makes it impossible to mutably manage linger even if `mutableUsers` is set.

We could make it contingent on `mutableUsers = false`, or we could make the default value `null` rather than `false` and only enable/disable-linger for users where it's explicitly set -- I think the latter would make the activation script a bit more complicated, but I prefer it.

**Comment by dschrempf** [comment]
Hm, maybe I missed something but if I do not persist the folder `/var/lib/systemd/linger`, lingering does not persist either on my setup. As far as I understand your replies, this behavior is unexpected because activation scripts are run during boot.

If you want, I can do more tests or create a separate issue.

**Comment by bamhm182** [comment]
> Hm, maybe I missed something but if I do not persist the folder `/var/lib/systemd/linger`, lingering does not persist either on my setup. As far as I understand your replies, this behavior is unexpected because activation scripts are run during boot.
> 
> If you want, I can do more tests or create a separate issue.

I noticed this as well. I created a tmpfiles rule to automatically create my `/var/lib/systems/linger/$user` file, but it seems like that rule should already be in there. Is there any reason we would expect linger to be set to true and that file should not exist?

**Comment by arjan-s** [comment]
In case anyone is still monitoring this, I found out that user timers aren't started even with `linger` enabled. That seems to be in contrast to the explanations above. Anyone here encountered this too?

**Comment by dschrempf** [comment]
For me, user-space timers work as expected even on servers. I am using `  users.users.USERNAME.linger = true;`.

If you are using impermanence, you may want to persist
      "/var/lib/systemd/linger"
      "/var/lib/systemd/timers"

**Comment by ppenguin** [comment]
@arjan-s @dschrempf 
Now that you mention: I have a user timer for changing wallpapers, and indeed it doesn't start after reboot.
I don't have linger on those hosts, since these are not servers and indeed I _don't_ need user services that survive login for them, but I _do_ expect timers to start automatically. 

In other words, it seems counter-intuitive to have user-timers depend on linger, but I guess `startServices` would be acceptable if it works as expected; does it after boot/login though?

**Comment by arjan-s** [comment]
@dschrempf Thank you! `systemd.user.startServices` was indeed the missing link! Even though I'm using impermanence, I didn't have to persist those two directories.

@ppenguin On my laptops, user timers and users services are autostarting and working perfectly, without `linger` and `startServices`. It sounds like your login method doesn't start the systemd user session. For some methods you need to enable this explicitly.

**Comment by thefossguy** [comment]
@ppenguin what does `systemctl status [--user] $your-timer.timer` show?

**Comment by crasm** [comment]
I was bit by lingering being default off when trying to use `systemd.user.tmpfiles.rules`. Trying to debug why the files weren't being created led me to note that `systemctl --user` failed with `Failed to connect to user scope bus via local transport: No such file or directory`. Enabling linger is the fix.
